core/tests/test-dependencies/root-files-mapping-generator.ts

```diff
@@ -0,0 +1,736 @@
+      }
+      return true;
+    });
+  }
+
+  /**
+   * Finds the root dependencies for the given file.
+   */
+  private getRootFilesOfFile(
+    file: string,
+    cache: Record<string, string[]> = {},
+    ignoreModules: boolean = true,
+    visited: Set<string> = new Set()
+  ): string[] {
+    if (cache[file]) {
+      return cache[file];
+    }
+    if (visited.has(file)) {
+      return [];
+    }
+    visited.add(file);
+
+    let references = this.getFilesWithDependency(file, ignoreModules);
+    if (references.length === 0 || this.pageModules.includes(file)) {
+      return [file];
+    }
+
+    const roots: string[] = [];
+    for (const reference of references) {
+      roots.push(
+        ...this.getRootFilesOfFile(reference, cache, ignoreModules, visited)
+      );
+    }
+
+    return Array.from(new Set(roots));
+  }
+
+  /**
+   * Gets the modules that correspond with the test suites.
+   */
+  private getTestSuiteModules(): string[] {
+    const testSuiteModules: string[] = [];
+    const testSuiteConfigFiles = fs
+      .readdirSync(CI_TEST_SUITE_CONFIGS_DIRECTORY)
+      .filter((file: string) => file.endsWith('.json'))
+      .map(file => path.join(CI_TEST_SUITE_CONFIGS_DIRECTORY, file));
+    for (const testSuiteConfig of testSuiteConfigFiles) {
+      const config = JSON.parse(fs.readFileSync(testSuiteConfig, 'utf-8'));
+      const suites = config.suites;
+      for (const suite of suites) {
+        testSuiteModules.push(suite.module);
+      }
+    }
+    return testSuiteModules;
+  }
+
+  /**
+   * Gets the valid root files.
+   */
+  private getValidRootFiles(): string[] {
+    const validRootFiles: string[] = [
+      ...this.pageModulesCoveredByRouteMapping,
+      ...this.getTestSuiteModules(),
+    ];
+    const rootFilesConfig = JSON.parse(
+      fs.readFileSync(ROOT_FILES_CONFIG_FILE, 'utf-8')
+    );
+    validRootFiles.push(...rootFilesConfig.VALID_ROOT_FILES);
+    validRootFiles.push(...rootFilesConfig.RUN_ALL_TESTS_ROOT_FILES);
+    return validRootFiles;
+  }
+
+  /**
+   * Validates the root files mapping.
+   */
+  private validateRootFilesMapping(
+    rootFilesMapping: Record<string, string[]>
+  ): void {
+    const rootFiles = Array.from(
+      new Set(Object.values(rootFilesMapping).flat())
+    );
+    const validRootFiles = this.getValidRootFiles();
+    const invalidRootFiles = rootFiles.filter((rootFile: string) => {
+      if (isFrontendTestFile(rootFile)) {
+        return false;
+      }
+      return !validRootFiles.includes(rootFile);
+    });
+    if (invalidRootFiles.length > 0) {
+      throw new Error(
+        'The following invalid root files were found when generating ' +
+          `the root files mapping:\n${invalidRootFiles.join('\n')}.\n` +
+          'Please add them to the VALID_ROOT_FILES or RUN_ALL_TESTS_ROOT_FILES ' +
+          'in the root files config at core/tests/root-files-config.json or ' +
+          'ensure that they are a valid test suite module or page module.'
```

Comment:
It would probably also make sense to add a link to a wiki page that would explain a bit more (for example what is the difference between VALID_ROOT_FILES or RUN_ALL_TESTS_ROOT_FILES. You can just create the wiki page now and fill it later.

Reply:
Is it fine to do this after, I am planning to do all WIKI pages after the other PRs, will make a main PR which will introduce these changes to the files, or I can do now (do I have to make a PR to introduce a new page on the repo?)

Reply:
Yep, please make the wiki now (using PR) and then fill it later.

Reply:
Ok created a PR here: https://github.com/oppia/oppia-web-developer-docs/pull/328https://github.com/oppia/oppia-web-developer-docs/pull/328